File: f77/blaio.f
C -*- Mode:Fortran; Coding:us-ascii-unix; fill-column:72 -*-
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC.H.S.CC
C @file      blaio.f
C @author    Mitch Richling http://www.mitchr.me/
c @brief     matrix/vector printing for examples. @EOL
C @keywords  blas lapack numerical linear algebra matrix vector netlib
C @std       F77 MIL-STD-1753
C @see       https://github.com/richmit/ex-blas-lapack/
C @copyright
C  @parblock
C  Copyright (c) 1995,2025, Mitchell Jay Richling All rights reserved.
C
C  Redistribution and use in source and binary forms, with or without
C  modification, are permitted provided that the following conditions
C  are met:
C
C  1. Redistributions of source code must retain the above copyright
C     notice, this list of conditions, and the following disclaimer.
C
C  2. Redistributions in binary form must reproduce the above
C     copyright notice, this list of conditions, and the following
C     disclaimer in the documentation and/or other materials provided
C     with the distribution.
C
C  3. Neither the name of the copyright holder nor the names of its
C     contributors may be used to endorse or promote products derived
C     from this software without specific prior written permission.
C
C  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
C  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
C  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
C  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
C  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
C  INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
C  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
C  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
C  HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
C  STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
C  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
C  OF THE POSSIBILITY OF SUCH DAMAGE.
C  @endparblock
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC.H.E.CC

C-----------------------------------------------------------------------
      subroutine sgeprt(m, n, a, c)
        integer m, n
        real a(m,n)
        character*(*) c
        call sprtmx(m, n, a, 9, 3, '[', ']', '[', ']', c)
      end

C-----------------------------------------------------------------------
      subroutine sprtmx(m, n, a, wide, prec, ldel,
     *                  rdel, lidel, ridel, tag)
      integer m, n, wide, prec
      real a(m, n)
      character*(*) ldel, rdel, lidel, ridel, tag
C     Local vars
      character fst*70
      integer i, j
C     Build format string
      write (fst, '("(1x,a",i0,",a,a,",i0,"f",i0,".",i0,",1x,a,a)")')
     * len(tag), n, wide, prec
C     Use format string and print out the matrix
      do i=1,m
         if( (i .eq. 1) .and. (i .eq. m) ) then
            write (*, fst) tag,ldel,lidel, (a(i, j), j=1,n), ridel,rdel
         else if(i .eq. 1) then
            write (*, fst) tag,ldel,lidel, (a(i, j), j=1,n), ridel,' '
         else if(i .eq. m) then
            write (*, fst) ' ',' ',lidel, (a(i, j), j=1,n),ridel,rdel
         else
            write (*, fst) ' ',' ',lidel, (a(i, j), j=1,n),ridel,' '
         endif
      enddo
      end

C-----------------------------------------------------------------------
      subroutine dgeprt(m, n, a, c)
        integer m, n
        double precision a(m,n)
        character*(*) c
        call dprtmx(m, n, a, 9, 3, '[', ']', '[', ']', c)
      end

C-----------------------------------------------------------------------
      subroutine dprtmx(m, n, a, wide, prec, ldel,
     *                  rdel, lidel, ridel, tag)
      integer m, n, wide, prec
      double precision  a(m, n)
      character*(*) ldel, rdel, lidel, ridel, tag
C     Local vars
      character fst*70
      integer i, j
C     Build format string
      write (fst, '("(1x,a",i0,",a,a,",i0,"f",i0,".",i0,",1x,a,a)")')
     * len(tag), n, wide, prec
C     Use format string and print out the matrix
      do i=1,m
         if( (i .eq. 1) .and. (i .eq. m) ) then
            write (*, fst) tag,ldel,lidel, (a(i, j), j=1,n), ridel,rdel
         else if(i .eq. 1) then
            write (*, fst) tag,ldel,lidel, (a(i, j), j=1,n), ridel,' '
         else if(i .eq. m) then
            write (*, fst) ' ',' ',lidel, (a(i, j), j=1,n),ridel,rdel
         else
            write (*, fst) ' ',' ',lidel, (a(i, j), j=1,n),ridel,' '
         endif
      enddo
      end

File: c/blaio.c
/* -*- Mode:C; Coding:us-ascii-unix; fill-column:158 -*- */
/*******************************************************************************************************************************************************.H.S.**/
/**
 @file      blaio.c
 @author    Mitch Richling http://www.mitchr.me/
 @brief     matrix/vector I/O.@EOL
 @keywords  blas numerical linear algebra matrix vector
 @std       C99
 @see       https://github.com/richmit/ex-blas-lapack/
 @copyright 
  @parblock
  Copyright (c) 1993,2025, Mitchell Jay Richling <http://www.mitchr.me/> All rights reserved.
  
  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
  
  1. Redistributions of source code must retain the above copyright notice, this list of conditions, and the following disclaimer.
  
  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions, and the following disclaimer in the documentation
     and/or other materials provided with the distribution.
  
  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software
     without specific prior written permission.
  
  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
  OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
  DAMAGE.
  @endparblock
 @filedetails   

  A few simple matrix/vector print/read functions useful for printing small test cases and debugging
*/
/*******************************************************************************************************************************************************.H.E.**/

/*------------------------------------------------------------------------------------------------------------------------------------------------------------*/
#include <stdlib.h>                                                      /* Standard Lib            C89      */
#include <stdio.h>                                                       /* I/O lib                 C89      */
#include <ctype.h>                                                       /* Char classes            C89      */
#include <string.h>                                                      /* Strings                 C89      */

/*------------------------------------------------------------------------------------------------------------------------------------------------------------*/
#include "blaio.h"                                                       /* Basic Linear Algebra I/O         */

/*------------------------------------------------------------------------------------------------------------------------------------------------------------*/
void printMatrixUbr(const enum CBLAS_ORDER order, /* CBLAS row order                                                 */
                    int n, int m, double *a,      /* Size and array                                                  */
                    char *inStr, char *outStr,    /* "in" string, and "out" string                                   */
                    double minIn, double maxIn,   /* Min/Max values for "in" range.                                  */
                    int wide, int prec,           /* Width and precesion for floats                                  */
                    int *rowPerm, int *colPerm,                                                                      
                    char prtMode,                 /* b=bitmap, V=values, *=in/out)                                   */
                    char *fileName, 
                    int maskMode,                 /* L=below diag, U=above diag, D=Diagnal, M=Mask, 0=NONE, \0=NONE  */
                    char *mask,
                    char *pad,                    /* Right pad string                                                */
                    char *ldel, char *rdel,       /* Left and right delimiter                                        */
                    char *lidel, char *ridel,     /* Left and right INNER delimiter                                  */
                    char *tag                     /* Tag for first line                                              */
  );

/*------------------------------------------------------------------------------------------------------------------------------------------------------------*/
void printVector(int n, double *v, int wide, int prec, char *pad, char *ldel, char *rdel, char *tag) {
  printMatrixUbr(CblasRowMajor, 1, n, v, NULL, NULL, 0.0, 0.0, wide, prec, NULL, NULL, 'V', NULL, '0', NULL, pad, ldel, rdel, "", "", tag);
} /* end func printVector */

/*------------------------------------------------------------------------------------------------------------------------------------------------------------*/
void printMatrix(const enum CBLAS_ORDER order, int n, int m, double *a, int wide, int prec, char *pad, char *ldel, char *rdel, char *lidel, char *ridel, char *tag) {
  printMatrixUbr(order, n, m, a, NULL, NULL, 0.0, 0.0, wide, prec, NULL, NULL, 'V', NULL, '0', NULL, pad, ldel, rdel, lidel, ridel, tag);
} /* end func printMatrix */

/*------------------------------------------------------------------------------------------------------------------------------------------------------------*/
void printMatrixThr(const enum CBLAS_ORDER order, int n, int m, double *a, char *inStr, char *outStr, double minIn, double maxIn, char *pad, char *ldel, char *rdel, char *lidel, char *ridel, char *tag) {
  printMatrixUbr(order, n, m, a, inStr, outStr, minIn, maxIn, 0, 0, NULL, NULL, '*', NULL, '0', NULL, pad, ldel, rdel, lidel, ridel, tag);
} /* end func printMatrixThr */

/*------------------------------------------------------------------------------------------------------------------------------------------------------------*/
void printMatrixUbr(const enum CBLAS_ORDER order, /* CBLAS row order                  */
                    int n, int m, double *a,      /* Size and array                   */
                    char *inStr, char *outStr,    /* "in" string, and "out" string    */
                    double minIn, double maxIn,   /* Min/Max values for "in" range.   */
                    int wide, int prec,           /* Width and precesion for floats   */
                    int *rowPerm, int *colPerm,   /* Permute rows i->xx[i]            */
                    char prtMode,                 /* b=bitmap, V=values, *=in/out     */
                    char *fileName,               /* if NULL, stdout.                 */
                    int maskMode,                 /* L, U, D, M-Mask, 0=NONE          */
                    char *mask,                   /* Mask (same size as a) ctrl print */
                    char *pad,                    /* Right pad string                 */
                    char *ldel, char *rdel,       /* Left and right delimiter         */
                    char *lidel, char *ridel,     /* Left and right INNER delimiter   */
                    char *tag                     /* Tag for first line               */
  ) {
  int i, j, iP, jP;
  int k, ldelLen, tagLen, cIdx, prtPerMask;
  double pVal;

  if(inStr == NULL)  inStr  = "*";
  if(outStr == NULL) outStr = " ";
  if(wide < 0)       wide   = 5;
  if(prec < 0)       prec   = 2;
  if(ldel == NULL)   ldel   = "[";
  if(ridel == NULL)  ridel  = "]";
  if(lidel == NULL)  lidel  = "[";
  if(rdel == NULL)   rdel   = "]";
  if(pad  == NULL)   pad    = " ";
  if(tag  == NULL)   tag    = "";

  ldelLen = strlen(ldel);
  tagLen = strlen(tag);
  for(j=0; j<n; j++) {
    if(j==0) 
      printf("%s%s%s%s", tag, ldel, lidel, pad);
    else {
      for(k=0;k<tagLen;k++) printf(" ");
      for(k=0;k<ldelLen;k++) printf(" ");
      printf("%s%s", lidel, pad);
    } /* end if/else */
    for(i=0; i<m; i++) {
      if(colPerm != NULL)
        iP = colPerm[i];
      else
        iP = i;
      if(rowPerm != NULL)
        jP = rowPerm[j];
      else
        jP = j;
      if(order == CblasColMajor)
        cIdx = n*iP+jP;
      else
        cIdx = m*jP+iP;
      pVal = a[cIdx];
      // Figure out what the mask has to do with printing..
      if(maskMode == '0')
        prtPerMask = 1;
      else if(maskMode == 'L')
        prtPerMask = (iP<jP);  // Row order specific!  Fix this.
      else if(maskMode == 'D')
        prtPerMask = (iP==jP);
      else if(maskMode == 'U')
        prtPerMask = (iP>jP);  // Row order specific!  Fix this.
      else if(maskMode == 'M')
        prtPerMask = mask[cIdx];
      else
        prtPerMask = 1;

      if(prtMode == '*') {
        if( prtPerMask && (pVal >= minIn) && (pVal <= maxIn) )
          printf("%s%s", inStr, pad);
        else 
          printf("%s%s", outStr, pad);
      } else {
        if(prtPerMask)
          printf("%*.*f%s", wide, prec, pVal, pad);
        else
          printf("%*s%s", wide, outStr, pad);
      } /* end if/else */
    } /* end for */
    if(j==n-1)
      printf("%s%s\n", ridel, rdel);
    else
      printf("%s\n", ridel);
  } /* end for */
} /* end func printMatrixUbr*/
